src/engine/audio/index.ts

```diff
@@ -0,0 +1,349 @@
+
+  const mainGain = new GainNode(context);
+  const mainBus = new ChannelMergerNode(context);
+  mainGain.connect(context.destination);
+  // mainBus.connect(mainGain);
+
+  const sampleBus = new ChannelMergerNode(context);
+  const sampleGain = new GainNode(context);
+  // sampleBus.connect(sampleGain);
+  // sampleGain.connect(mainBus);
+  sampleGain.connect(mainGain);
+
+  const sampleCache = new Map<string, AudioBuffer>();
+
+  const tripleBuffer = createTripleBuffer();
+  const cursorBuffers = tripleBuffer.buffers.map((b) => createCursorBuffer(b));
+  const transformViews = cursorBuffers.map(
+    (buffer) =>
+      ({
+        // note: needs synced with renderableBuffer properties in game worker
+        // todo: abstract the need to sync structure with renderableBuffer properties
+        worldMatrix: addViewMatrix4(buffer, maxEntities),
+        worldMatrixNeedsUpdate: addView(buffer, Uint8Array, maxEntities),
+      } as TransformView)
+  );
+
+  const entityPanners = new Map<number, PannerNode>();
+
+  const sampleQueue: [string, number][] = [];
+
+  return {
+    context,
+    tripleBuffer,
+    transformViews,
+    entityPanners,
+    listenerEntity: NOOP,
+    peerEntities: new Map(),
+    main: {
+      bus: mainBus,
+      gain: mainGain,
+    },
+    sample: {
+      bus: sampleBus,
+      gain: sampleGain,
+      cache: sampleCache,
+      queue: sampleQueue,
+    },
+  };
+};
+
+export const preloadDefaultAudio = async (
+  audioState: AudioState,
+  defaultAudioFiles: string[] = ["/audio/cricket.ogg"]
+) => {
+  defaultAudioFiles.forEach(async (file) => await getAudioBuffer(audioState, file));
+};
+
+export const setAudioListener = (audioState: AudioState, eid: number) => {
+  audioState.listenerEntity = eid;
+};
+
+export const setAudioPeerEntity = (audioState: AudioState, peerId: string, eid: number) => {
+  audioState.peerEntities.set(peerId, eid);
+};
+
+export const bindAudioStateEvents = (audioState: AudioState, gameWorker: Worker) => {
+  // todo: register messages on a single event listener
+  gameWorker.addEventListener("message", ({ data }) => {
+    if (typeof data !== "object") {
+      return;
+    }
+
+    const message = data as WorkerMessages;
+
+    switch (message.type) {
+      case WorkerMessageType.PlayAudio:
+        playAudio(audioState, message.filepath, message.eid);
+        break;
+      case WorkerMessageType.SetAudioListener:
+        setAudioListener(audioState, message.eid);
+        break;
+      case WorkerMessageType.SetAudioPeerEntity:
+        setAudioPeerEntity(audioState, message.peerId, message.eid);
+        break;
+    }
+  });
+};
+
+export function initAudioState(gameWorker: Worker) {
+  const audioState = createAudioState();
+  preloadDefaultAudio(audioState);
+  bindAudioStateEvents(audioState, gameWorker);
+  return audioState;
+}
+
+export const disposeAudioState = (audioState: AudioState) => {
+  audioState.context.close();
+};
+
+/* Main Thread Systems */
```

Comment:
I think we'll want to split this file into multiple modules with prefixes that indicate what thread they are running on.